File: tests/algorithms/x/PHP/ciphers/deterministic_miller_rabin.php
<?php
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
function mod_pow($base, $exp, $mod) {
  $result = 1;
  $b = $base % $mod;
  $e = $exp;
  while ($e > 0) {
  if ($e % 2 == 1) {
  $result = ($result * $b) % $mod;
}
  $b = ($b * $b) % $mod;
  $e = _intdiv($e, 2);
};
  return $result;
}
function miller_rabin($n, $allow_probable) {
  if ($n == 2) {
  return true;
}
  if ($n < 2 || $n % 2 == 0) {
  return false;
}
  if ($n > 5) {
  $last = $n % 10;
  if (!($last == 1 || $last == 3 || $last == 7 || $last == 9)) {
  return false;
};
}
  $limit = 3825123056546413051;
  if ($n > $limit && (!$allow_probable)) {
  $panic('Warning: upper bound of deterministic test is exceeded. Pass allow_probable=true to allow probabilistic test.');
}
  $bounds = [2047, 1373653, 25326001, 3215031751, 2152302898747, 3474749660383, 341550071728321, $limit];
  $primes = [2, 3, 5, 7, 11, 13, 17, 19];
  $i = 0;
  $plist_len = count($primes);
  while ($i < count($bounds)) {
  if ($n < $bounds[$i]) {
  $plist_len = $i + 1;
  $i = count($bounds);
} else {
  $i = $i + 1;
}
};
  $d = $n - 1;
  $s = 0;
  while ($d % 2 == 0) {
  $d = _intdiv($d, 2);
  $s = $s + 1;
};
  $j = 0;
  while ($j < $plist_len) {
  $prime = $primes[$j];
  $x = mod_pow($prime, $d, $n);
  $pr = false;
  if ($x == 1 || $x == $n - 1) {
  $pr = true;
} else {
  $r = 1;
  while ($r < $s && (!$pr)) {
  $x = ($x * $x) % $n;
  if ($x == $n - 1) {
  $pr = true;
}
  $r = $r + 1;
};
}
  if (!$pr) {
  return false;
}
  $j = $j + 1;
};
  return true;
}
echo rtrim(_str(miller_rabin(561, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(563, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(838201, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(838207, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(17316001, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(17316017, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(3078386641, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(3078386653, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(1713045574801, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(1713045574819, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(2779799728307, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(2779799728327, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(113850023909441, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(113850023909527, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(1275041018848804351, false))), PHP_EOL;
echo rtrim(_str(miller_rabin(1275041018848804391, false))), PHP_EOL;

File: tests/algorithms/transpiler/FS/ciphers/deterministic_miller_rabin.fs
// Generated 2025-08-07 08:16 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec mod_pow (``base``: int) (exp: int) (``mod``: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    let mutable ``mod`` = ``mod``
    try
        let mutable result: int = 1
        let mutable b: int = ((``base`` % ``mod`` + ``mod``) % ``mod``)
        let mutable e: int = exp
        while e > 0 do
            if (((e % 2 + 2) % 2)) = 1 then
                result <- (((result * b) % ``mod`` + ``mod``) % ``mod``)
            b <- (((b * b) % ``mod`` + ``mod``) % ``mod``)
            e <- e / 2
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
let rec miller_rabin (n: int) (allow_probable: bool) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable n = n
    let mutable allow_probable = allow_probable
    try
        if n = 2 then
            __ret <- true
            raise Return
        if (n < 2) || ((((n % 2 + 2) % 2)) = 0) then
            __ret <- false
            raise Return
        if n > 5 then
            let last: int = ((n % 10 + 10) % 10)
            if not ((((last = 1) || (last = 3)) || (last = 7)) || (last = 9)) then
                __ret <- false
                raise Return
        let limit: int64 = 3825123056546413051L
        if ((int64 n) > limit) && (not allow_probable) then
            failwith ("Warning: upper bound of deterministic test is exceeded. Pass allow_probable=true to allow probabilistic test.")
        let bounds: int64 array = [|int64 (2047); int64 (1373653); int64 (25326001); 3215031751L; 2152302898747L; 3474749660383L; 341550071728321L; limit|]
        let primes: int array = [|2; 3; 5; 7; 11; 13; 17; 19|]
        let mutable i: int = 0
        let mutable plist_len: int = Seq.length (primes)
        while i < (Seq.length (bounds)) do
            if (int64 n) < (_idx bounds (i)) then
                plist_len <- i + 1
                i <- Seq.length (bounds)
            else
                i <- i + 1
        let mutable d: int = n - 1
        let mutable s: int = 0
        while (((d % 2 + 2) % 2)) = 0 do
            d <- d / 2
            s <- s + 1
        let mutable j: int = 0
        while j < plist_len do
            let prime: int = _idx primes (j)
            let mutable x: int = mod_pow (prime) (d) (n)
            let mutable pr: bool = false
            if (x = 1) || (x = (n - 1)) then
                pr <- true
            else
                let mutable r: int = 1
                while (r < s) && (not pr) do
                    x <- (((x * x) % n + n) % n)
                    if x = (n - 1) then
                        pr <- true
                    r <- r + 1
            if not pr then
                __ret <- false
                raise Return
            j <- j + 1
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
printfn "%s" (_str (miller_rabin (561) (false)))
printfn "%s" (_str (miller_rabin (563) (false)))
printfn "%s" (_str (miller_rabin (838201) (false)))
printfn "%s" (_str (miller_rabin (838207) (false)))
printfn "%s" (_str (miller_rabin (17316001) (false)))
printfn "%s" (_str (miller_rabin (17316017) (false)))
printfn "%s" (_str (miller_rabin (int 3078386641L) (false)))
printfn "%s" (_str (miller_rabin (int 3078386653L) (false)))
printfn "%s" (_str (miller_rabin (int 1713045574801L) (false)))
printfn "%s" (_str (miller_rabin (int 1713045574819L) (false)))
printfn "%s" (_str (miller_rabin (int 2779799728307L) (false)))
printfn "%s" (_str (miller_rabin (int 2779799728327L) (false)))
printfn "%s" (_str (miller_rabin (int 113850023909441L) (false)))
printfn "%s" (_str (miller_rabin (int 113850023909527L) (false)))
printfn "%s" (_str (miller_rabin (int 1275041018848804351L) (false)))
printfn "%s" (_str (miller_rabin (int 1275041018848804391L) (false)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
